File: Problems/No.1617_Count_Subtrees_With_Max_Distance_Between_Cities/count_subtrees_with_max_distance_between_cities.py
class Solution(object):
    def countSubgraphsForEachDiameter(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: List[int]
        """

        # parameters
        dis_cnt = [0] * (n-1)
        edge_info = [[] for _ in range(n)]

        # start computing
        # construct edge information
        for [city_1, city_2] in edges:
            edge_info[city_1-1].append(city_2)
            edge_info[city_2-1].append(city_1)
        
        # check all possible subsets
        dis_cnt[0] = len(edges)
        for i in range(2**n):
            # decide current subset and check if it's size >= 3
            curr_subset = []
            for j in range(n):
                if ((i & (1<<j)) != 0):
                    curr_subset.append(j+1)
            if (len(curr_subset) < 3):
                continue
            
            # calculate the max distance of this subset
            MAX_distance = 0
            for city in curr_subset:
                curr_distance = -1
                subset_copy = curr_subset[:]

                # visit from each city in the current subset using BFS
                visit_city = []
                visit_city.append(city)
                subset_copy.remove(city)
                while visit_city:
                    curr_distance += 1
                    Q_size = len(visit_city)
                    for cnt in range(Q_size):
                        curr_city = visit_city.pop(0)
                        for next_city in edge_info[curr_city-1]:
                            if next_city in subset_copy:
                                visit_city.append(next_city)
                                subset_copy.remove(next_city)
                
                # check if every cities in the subset can be visited
                if subset_copy:
                    break
                else:
                    MAX_distance = max(MAX_distance, curr_distance)
            
            if (MAX_distance > 1):
                dis_cnt[MAX_distance-1] += 1

        return dis_cnt

File: Problems/No.1617_Count_Subtrees_With_Max_Distance_Between_Cities/count_subtrees_with_max_distance_between_cities.cpp
class Solution {
public:
    vector<int> countSubgraphsForEachDiameter(int n, vector<vector<int>>& edges) {
        
        // parameters
        std::vector<int> dis_cnt(n-1, 0);
        std::unordered_map<int, std::vector<int>> edge_info;

        // start computing
        // construct edge information
        for (auto &edge: edges) {

            int city_1 = edge[0], city_2 = edge[1];
            edge_info[city_1].push_back(city_2);
            edge_info[city_2].push_back(city_1);

        }
        
        // check all possible enumerations
        dis_cnt[0] = edges.size();
        for (int i=0; i<std::pow(2, n); i++) {

            // decide current subset and check if it's size > 1
            std::unordered_set<int> curr_subset;
            for (int j=0; j<n; j++)
                if ((i & (1<<j)) != 0) curr_subset.insert(j+1);
            if (curr_subset.size() < 3) continue;

            // calculate the max distance of this subset
            int MAX_distance = 0;
            for (auto &city: curr_subset) {
                
                int curr_distance = -1;
                auto subset_copy = curr_subset;

                // visit from each city in the current subset using queue
                std::queue<int> visit_city;
                visit_city.push(city);
                subset_copy.erase(city);
                while (!visit_city.empty()) {

                    curr_distance++;
                    int Q_size = visit_city.size();
                    for (int cnt=0; cnt<Q_size; cnt++) {

                        int curr_city = visit_city.front();
                        visit_city.pop();
                        for (auto &next_city: edge_info[curr_city]) 
                            if (subset_copy.count(next_city)) {

                                visit_city.push(next_city);
                                subset_copy.erase(next_city);

                            }

                    }

                }

                // check if every cities in the subset can be visited
                if (!subset_copy.empty()) break;
                else MAX_distance = std::max(MAX_distance, curr_distance);
                
            }
            if (MAX_distance > 1) dis_cnt[MAX_distance-1]++;

        }

        return dis_cnt;

    }
};
